src/value_serializer.rs

```diff
@@ -0,0 +1,382 @@
+    assert_eq!(data_ptr, embedder_ptr);
+    vtable
+  }
+
+  pub fn new<T>() -> Self
+  where
+    T: ValueSerializerDelegateImpl,
+  {
+    Self {
+      cxx_base: Self::construct_cxx_base(),
+      offset_within_embedder: Self::get_offset_within_embedder::<T>(),
+      rust_vtable: Self::get_rust_vtable::<T>(),
+    }
+  }
+
+  pub unsafe fn dispatch(
+    value_serializer_delegate: &ValueSerializerDelegate,
+  ) -> &dyn ValueSerializerDelegateImpl {
+    let this = Self::get_cxx_base_offset()
+      .to_embedder::<Self>(value_serializer_delegate);
+    let embedder = this.offset_within_embedder.to_embedder::<Opaque>(this);
+    std::mem::transmute((embedder, this.rust_vtable))
+  }
+
+  pub unsafe fn dispatch_mut(
+    value_serializer_delegate: &mut ValueSerializerDelegate,
+  ) -> &mut dyn ValueSerializerDelegateImpl {
+    let this = Self::get_cxx_base_offset()
+      .to_embedder_mut::<Self>(value_serializer_delegate);
+    let vtable = this.rust_vtable;
+    let embedder = this.offset_within_embedder.to_embedder_mut::<Opaque>(this);
+    std::mem::transmute((embedder, vtable))
+  }
+
+  pub unsafe fn dispatch_box(
+    value_serializer_delegate: &mut ValueSerializerDelegate,
+  ) -> Box<dyn ValueSerializerDelegateImpl> {
+    std::mem::transmute(Self::dispatch_mut(value_serializer_delegate))
+  }
+}
+
+// ValueSerializer
+
+// Must be == sizeof(v8::ValueSerializer), see v8__ValueSerializer__CONSTRUCT().
+#[repr(C)]
+pub struct ValueSerializer([usize; 1]);
+
+extern "C" {
+  fn v8__ValueSerializer__CONSTRUCT(
+    buf: *mut MaybeUninit<ValueSerializer>,
+    isolate: *mut Isolate,
+  );
+  fn v8__ValueSerializer__CONSTRUCT_with_delegate(
+    buf: *mut MaybeUninit<ValueSerializer>,
+    isolate: *mut Isolate,
+    delegate: *mut ValueSerializerDelegate,
+  );
+
+  fn v8__ValueSerializer__DESTRUCT(this: *mut ValueSerializer);
+
+  fn v8__ValueSerializer__Release(
+    this: *mut ValueSerializer,
+    ptr: *mut *mut u8,
+    size: *mut usize,
+  );
+
+  fn v8__ValueSerializer__TransferArrayBuffer(
+    this: *mut ValueSerializer,
+    transfer_id: u32,
+    array_buffer: Local<ArrayBuffer>,
+  );
+
+  fn v8__ValueSerializer__WriteHeader(this: *mut ValueSerializer);
+  fn v8__ValueSerializer__WriteValue(
+    this: *mut ValueSerializer,
+    context: Local<Context>,
+    value: Local<Value>,
+  ) -> MaybeBool;
+  fn v8__ValueSerializer__WriteUint32(this: *mut ValueSerializer, value: u32);
+  fn v8__ValueSerializer__WriteUint64(this: *mut ValueSerializer, value: u64);
+  fn v8__ValueSerializer__WriteDouble(this: *mut ValueSerializer, value: f64);
+  fn v8__ValueSerializer__WriteRawBytes(
+    this: *mut ValueSerializer,
+    source: *const c_void,
+    length: usize,
+  );
+}
+
+impl ValueSerializer {
+  pub fn new(scope: &mut HandleScope<'_>) -> Self {
+    let isolate: *mut Isolate = scope.get_isolate_ptr();
+    let mut value_serializer: MaybeUninit<Self> = MaybeUninit::uninit();
+    unsafe {
```

Comment:
With the current implementation where release() returns a `Box<[u8]>` we will have to provide a default `ValueSerializedDelegate`, because otherwise the default behavior will be to allocate memory with an unknown allocator (malloc()/free()) which does not necessarily produce a valid `Box<[u8]>`.